### PR TITLE
Merge 0.9.0 into eth-bridge-integration

### PR DESCRIPTION
**Pull request by juped**
wip, still needs rpc reimplemented

### CODE REVIEW

**Comment by james-chf** [comment]
pls update wasm

**Comment by james-chf** [comment]
Merging this PR should close https://github.com/anoma/namada/issues/730

**Comment by james-chf** [comment]
Current status: refactoring of bridge pool query endpoints to work with the new RPC router is WIP, some tests are still broken. There are still some conflicts with latest `eth-bridge-integration` that need to be merged in, after those tests are fixed up.

**Comment by james-chf** [comment]
Compiling with `abcipp` is also broken, though this is also the case on `main` - have reappropiated https://github.com/anoma/namada/issues/497 as an issue to fix this up in both the branches

**Comment by james-chf** [comment]
pls update wasm

File: shared/src/ledger/events.rs [hunk 1]
```diff
@@ -153,7 +162,7 @@ impl From<ProposalEvent> for Event {
 }
 
 /// Convert our custom event into the necessary tendermint proto type
-impl From<Event> for crate::facade::tendermint_proto::abci::Event {
+impl From<Event> for tendermint_proto::abci::Event {
```
Comment (sug0) [inline shared/src/ledger/events.rs:165]: this should be `crate::facade::tendermint_proto::abci::Event`, right?

**Review by sug0 — APPROVED** [review]

File: apps/src/lib/client/rpc.rs [hunk 2]
```diff
@@ -1369,100 +1320,77 @@ pub async fn query_storage_value<T>(
...
-                    "Error in the query {} (error code {})",
-                    response.info, err
-                )
-            }
-        }
+    // In case `T` is a unit (only thing that encodes to 0 bytes), we have to
+    // use `storage_has_key` instead of `storage_value`, because `storage_value`
+    // returns 0 bytes when the key is not found.
+    let maybe_unit = T::try_from_slice(&[]);
+    if let Ok(unit) = maybe_unit {
```
Comment (batconjurer) [inline apps/src/lib/client/rpc.rs:1327]: wow. what a hack :laughing:

File: shared/src/ledger/events.rs (hunk 1 above)
Comment (james-chf) [reply shared/src/ledger/events.rs:165]: There is no `facade` module in `shared` (currently)`, though there is something similar in `shared/lib.rs`

https://github.com/anoma/namada/blob/6cbce4a4dcdd3811679f5eb4aa7c82f51dff74e6/shared/src/lib.rs#L9-L16

Comment (sug0) [reply shared/src/ledger/events.rs:165]: oh right, this isn't in `apps`... yeah, we can use `crate::<dep>` inside of `shared/`

Comment (james-chf) [reply shared/src/ledger/events.rs:165]: Oh yeah, this should be using that, will update

File: shared/src/types/keccak.rs [hunk 3]
```diff
@@ -103,13 +104,48 @@ pub fn keccak_hash(bytes: &[u8]) -> KeccakHash {
 
 /// This module defines encoding methods compatible with Ethereum
 /// smart contracts.
+// TODO: relocate this sub-module to a new home
```
Comment (batconjurer) [inline shared/src/types/keccak.rs:107]: I've been thinking about this for a while too. Maybe we should just have an ABI mod which has both encode and keccak submods

File: shared/src/types/keccak.rs [hunk 4]
```diff
@@ -103,13 +104,48 @@ pub fn keccak_hash(bytes: &[u8]) -> KeccakHash {
...
+        encoded_data: Vec<u8>,
+        /// Indicate we do not own values of type `T`.
+        ///
+        /// Passing `PhantomData<T>` here would trigger the drop checker,
+        /// which is not the desired behavior, since we own an encoded value
+        /// of `T`, not a value of `T` itself.
+        _marker: PhantomData<*const T>,
+    }
+
+    impl<T> EncodeCell<T> {
```
Comment (batconjurer) [inline shared/src/types/keccak.rs:130]: This is interesting. Should we change the return type of the `Encode::encode` method to return an instance of this struct? @sug0 ?

File: tests/src/e2e/helpers.rs [hunk 5]
```diff
@@ -148,6 +149,63 @@ pub fn get_epoch(test: &Test, ledger_address: &str) -> Result<Epoch> {
     Ok(Epoch(epoch))
 }
 
+/// Get the last committed block height.
+pub fn get_height(test: &Test, ledger_address: &str) -> Result<u64> {
```
Comment (batconjurer) [inline tests/src/e2e/helpers.rs:153]: :frowning_face:

**Review by batconjurer — APPROVED** [review]
One question really.

File: shared/src/types/keccak.rs (hunk 4 above)
Comment (batconjurer) [reply shared/src/types/keccak.rs:130]: I like it.

**Pull request merged**